File: telemetry/src/main/java/org/strykeforce/thirdcoast/telemetry/item/UltrasonicRangefinderItem.java
package org.strykeforce.thirdcoast.telemetry.item;

import static org.strykeforce.thirdcoast.telemetry.grapher.Measure.VALUE;

import com.ctre.phoenix.CANifier;
import com.ctre.phoenix.CANifier.PWMChannel;
import java.util.Collections;
import java.util.EnumSet;
import java.util.Set;
import java.util.function.DoubleSupplier;
import org.strykeforce.thirdcoast.telemetry.grapher.Measure;

public class UltrasonicRangefinderItem extends AbstractItem {

  private static final Set<Measure> MEASURES = Collections.unmodifiableSet(EnumSet.of(VALUE));

  private final int deviceId;
  private final CANifier canifier;
  private final PWMChannel pwmChannel;
  private final double[] dutyCycleAndPeriod = new double[2];

  public UltrasonicRangefinderItem(int canId, PWMChannel pwmChannel) {
    super("sensor", "Ultrasonic Rangefinder (" + canId + ", " + pwmChannel + ")", MEASURES);
    this.pwmChannel = pwmChannel;
    deviceId = canId * 10 + pwmChannel.value;
    canifier = new CANifier(canId);
  }

  @Override
  public int deviceId() {
    return deviceId;
  }

  @Override
  public DoubleSupplier measurementFor(Measure measure) {
    if (!MEASURES.contains(measure)) {
      throw new IllegalArgumentException("invalid measure: " + measure.name());
    }
    return () -> {
      canifier.getPWMInput(pwmChannel, dutyCycleAndPeriod);
      return dutyCycleAndPeriod[0];
    };
  }
}

File: telemetry/src/main/kotlin/org/strykeforce/thirdcoast/telemetry/item/UltrasonicRangefinderItem.kt
package org.strykeforce.thirdcoast.telemetry.item

import com.ctre.phoenix.CANifier
import com.ctre.phoenix.CANifier.PWMChannel
import org.strykeforce.thirdcoast.telemetry.grapher.Measure
import org.strykeforce.thirdcoast.telemetry.grapher.Measure.VALUE
import java.util.*
import java.util.function.DoubleSupplier

class UltrasonicRangefinderItem(canId: Int, private val pwmChannel: PWMChannel) :
  AbstractItem("sensor", "Ultrasonic Rangefinder ($canId, $pwmChannel)", MEASURES) {

  private val deviceId: Int = canId * 10 + pwmChannel.value
  private val canifier: CANifier = CANifier(canId)
  private val dutyCycleAndPeriod = DoubleArray(2)

  override fun deviceId(): Int {
    return deviceId
  }

  override fun measurementFor(measure: Measure): DoubleSupplier {
    if (!MEASURES.contains(measure)) {
      throw IllegalArgumentException("invalid measure: " + measure.name)
    }
    return {
      canifier.getPWMInput(pwmChannel, dutyCycleAndPeriod)
      dutyCycleAndPeriod[0]
    } as DoubleSupplier
  }

  companion object {

    private val MEASURES = Collections.unmodifiableSet(EnumSet.of(VALUE))
  }
}
